cpp/src/replace/replace.cu

```diff
@@ -13,31 +16,528 @@
+      output_is_valid = input_is_valid;
+
+      bitmask = __ballot_sync(active_mask, input_is_valid);
+
+      if (input_is_valid) {
+        thrust::tie(output_data[i], output_is_valid)  =
+            get_new_value<T, replacement_has_nulls>(i, input_data,
+                                      values_to_replace_begin,
+                                      values_to_replace_end,
+                                      d_replacement_values,
+                                      replacement_valid);
+      }
+
+    } else {
+       thrust::tie(output_data[i], output_is_valid) =
+            get_new_value<T, replacement_has_nulls>(i, input_data,
+                                      values_to_replace_begin,
+                                      values_to_replace_end,
+                                      d_replacement_values,
+                                      replacement_valid);
+    }
+
+    /* output valid counts calculations*/
+    if (input_has_nulls or replacement_has_nulls){
+
+      bitmask &= __ballot_sync(active_mask, output_is_valid);
+
+      if(0 == (threadIdx.x % cudf::experimental::detail::warp_size)){
+        output_valid[(int)(i/cudf::experimental::detail::warp_size)] = bitmask;
+        valid_sum[(int)(threadIdx.x / cudf::experimental::detail::warp_size)] += __popc(bitmask);
+      }
+    }
+
+    i += blockDim.x * gridDim.x;
+    active_mask = __ballot_sync(active_mask, i < nrows);
+  }
+  if(input_has_nulls or replacement_has_nulls){
+    __syncthreads(); // waiting for the valid counts of each warp to be ready
+
+    // Compute total valid count for this block and add it to global count
+    uint32_t block_valid_count = sum_warps<uint32_t>(valid_sum);
+
+    // one thread computes and adds to output_valid_count
+    if (threadIdx.x < cudf::experimental::detail::warp_size && 0 == (threadIdx.x % cudf::experimental::detail::warp_size)) {
+      atomicAdd(output_valid_count, block_valid_count);
+    }
+  }
+}
+
+  /* --------------------------------------------------------------------------*/
+  /**
+   * @brief Functor called by the `type_dispatcher` in order to invoke and instantiate
+   *        `replace_kernel` with the appropriate data types.
+   */
+  /* ----------------------------------------------------------------------------*/
+  struct replace_kernel_forwarder {
+    template <typename col_type,
+              std::enable_if_t<cudf::is_fixed_width<col_type>()>* = nullptr>
+    void operator()(cudf::column_view const& input_col,
+                    cudf::column_view const& values_to_replace,
+                    cudf::column_view const& replacement_values,
+                    cudf::mutable_column_view& output,
+                    cudaStream_t stream = 0)
+    {
+      cudf::size_type *valid_count = nullptr;
+      rmm::device_scalar<cudf::size_type> valid_counter(0);
+      if (output.nullable()) {
+        valid_count = valid_counter.data();
```

Comment:
This doesn't really save anything to do this. I'd just pass `valid_counter.data()` to the kernel regardless.

Reply:
Done.